src/dune_rules/pkg_rules.ml

```diff
@@ -1335,22 +1335,29 @@ module DB = struct
     ;;
   end
 
-  type t =
-    { pkg_digest_table : Pkg_table.t
-    ; system_provided : Package.Name.Set.t
-    }
+  module T = struct
+    type t =
+      { pkg_digest_table : Pkg_table.t
+      ; system_provided : Package.Name.Set.t
+      }
 
-  let equal t ({ pkg_digest_table; system_provided } as t') =
-    phys_equal t t'
-    || (Pkg_table.equal t.pkg_digest_table pkg_digest_table
-        && Package.Name.Set.equal t.system_provided system_provided)
-  ;;
+    let equal t ({ pkg_digest_table; system_provided } as t') =
+      phys_equal t t'
+      || (Pkg_table.equal t.pkg_digest_table pkg_digest_table
+          && Package.Name.Set.equal t.system_provided system_provided)
+    ;;
 
-  let hash = `Do_not_hash
-  let _ = hash
-  (* Because t is large, hashing is expensive, so much so that hashing the db in Input.t
-     below slowed down the dune call in the test repo described in #12248 from 1s to
-     2s. *)
+    let hash { pkg_digest_table = _; system_provided = _ } = 0
```

Comment:
Instead of doing this, how about you just generate an id using Stdune.Id and return that? That way you won't need physical equality and you'll have a decent hash function.